### PR TITLE
Fix identical entrypoints get deduped

### PR DESCRIPTION
**Pull request by digitalsadhu**
## JIRA Issue
[COREWEB-120](https://jira.finn.no/browse/COREWEB-120)

## Status
**READY**

## Description
Refactor reader to unpack feeds and then process all unpacked directories with browserify.

After trying a number of other approaches we realised it would likely be better to stop using lower level browserify apis and just run browserify over the code and let it handle the edge cases.

## Todos
- [x] Tests

## Related Issues
* #59

### CODE REVIEW

File: test/reader.test.js [hunk 1]
```diff
@@ -41,17 +41,6 @@ test('should successfully bundle 2 feeds', async () => {
     expect(clean(prettier.format(content))).toMatchSnapshot();
 });
 
-test('should successfully bundle 2 feeds using automatic temp directory', async () => {
```
Comment (SimenB) [inline test/reader.test.js]: remove the code for automatic temp as well, and just require a passed in directory?

Comment (digitalsadhu) [reply test/reader.test.js]: It would be nice not to have to change the api or pass in what feels like an internal detail no?

Comment (SimenB) [reply test/reader.test.js]: We're unable to test it though, can break at any time.

API wise it's nicer though, so maybe we should investigate the CI failure?

Comment (digitalsadhu) [reply test/reader.test.js]: yea, probably should eh.

File: README.md [hunk 2]
```diff
@@ -4,38 +4,34 @@
...
-
-
+This takes one or more asset feeds, as produced by the [asset-pipe-js-writer][asset-pipe-js-writer], and produces an executable javascript bundle for the browser.
 
 ## Data format
 
-What we refere to as an asset feed is the internal data format used in [Browserify][browserify]. We
-use the exact same data format as Browserify in the [asset-pipe project][asset-pipe].
+What we refer to as asset feeds are arrays of objects in the internal data format used by [Browserify][browserify]. We
+use this exact same data format throughout the [asset-pipe project][asset-pipe].
```
Comment (SimenB) [inline README.md]: "data structure"?

File: README.md [hunk 3]
```diff
@@ -4,38 +4,34 @@
 
 This is an internal module intended for use by other modules in the [asset-pipe project][asset-pipe].
 
-This module can take one or multiple asset feeds, as produced by the [asset-pipe-js-writer][asset-pipe-js-writer], and produced an executable javascript bundle for the browser.
-
-
+This takes one or more asset feeds, as produced by the [asset-pipe-js-writer][asset-pipe-js-writer], and produces an executable javascript bundle for the browser.
```
Comment (SimenB) [inline README.md]: /s/javascript/JavaScript

File: lib/hash-feed.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * @param {object[]} feed - browserify compatible bundle feed
+ * @returns {Promise<string>} - hash of all id properties in given feed
+ */
+module.exports = function hashFeed(feed) {
+    const source = new Readable({ objectMode: true, read() {} });
+    const hasher = new IdHasher();
+
+    source.pipe(hasher);
+
+    return new Promise(resolve => {
```
Comment (SimenB) [inline lib/hash-feed.js]: needs error handling (`hasher.on('error', reject)`, or something)?

File: lib/hash-feed.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        hasher.on('finish', () => {
```
Comment (SimenB) [inline lib/hash-feed.js]: can't we just `await getStream(hasher)`?

File: lib/reader.js [hunk 6]
```diff
@@ -51,23 +30,22 @@ module.exports = async feeds => {
         'Expected at least 1 feed with more than one entry'
     );
 
-    const rows = getRows(feeds);
+    const opts = {
+        directory: await directory({ unsafeCleanup: true }),
```
Comment (SimenB) [inline lib/reader.js]: This will always create the tmp dir even if the consumer passes it in. Should check if it's passed before executing this function

File: lib/reader.js [hunk 7]
```diff
@@ -1,46 +1,25 @@
...
-module.exports = async feeds => {
+const feedEntrypoint = require('./feed-entrypoint');
+const unpackFeed = require('./unpack-feed');
+const { join } = require('path');
+const browserify = require('browserify');
+const getStream = require('get-stream');
+const hashFeed = require('./hash-feed');
+const tmp = require('tmp');
+const { promisify } = require('util');
+const directory = promisify(tmp.dir);
```
Comment (SimenB) [inline lib/reader.js]: Can use https://github.com/sindresorhus/tempy, since you're adding a dependency anyways

File: lib/reader.js [hunk 8]
```diff
@@ -51,23 +30,22 @@ module.exports = async feeds => {
...
-    await rows.reduce((acc, row) => {
-        acc.then(
-            () =>
-                new Promise(resolve => {
-                    browserifyPack.write(row, null, resolve);
-                })
-        );
-        return acc;
-    }, Promise.resolve());
+        await unpackFeed(root, feed);
```
Comment (SimenB) [inline lib/reader.js]: should we use `Promise.all` here to unpack in parallel?

File: lib/replace-dependencies.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+ * @example
+ * Given a depMap with entry lodash -> '/my/path'
+ * require('lodash') will be changed to require('/my/path')
+ *
+ * @param {string} root - root path to be prepended to dep map values
+ * @param {Map} depMap - map dependency name -> path
+ * @param {object} item - object with key source containing javascript source code
+ * @returns {Promise<string>} - source code with dependencies replaced
+ */
+module.exports = async function replaceDependencies(root, depMap, item) {
```
Comment (SimenB) [inline lib/replace-dependencies.js]: not `async`

File: package.json [hunk 10]
```diff
@@ -23,11 +26,18 @@
   },
   "license": "MIT",
   "dependencies": {
+    "asset-pipe-common": "^1.0.0-beta.6",
```
Comment (SimenB) [inline package.json]: we should scope and release 1.0.0 of this

File: package.json [hunk 11]
```diff
@@ -60,10 +70,14 @@
...
   },
-  "files": ["lib"],
+  "files": [
+    "lib"
+  ],
   "jest": {
     "clearMocks": true,
-    "coveragePathIgnorePatterns": ["mock"],
+    "coveragePathIgnorePatterns": [
+      "mock"
```
Comment (SimenB) [inline package.json]: Please do `"<rootDir>/mock/` or something to be more explicit

File: test/feed-entrypoint.test.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        { id: 1, file: '', source: '', deps: {} },
+        { id: 2, file: '', source: '', deps: {} },
+        { id: 3, file: '', source: '', deps: {}, entry: true },
+        { id: 4, file: '', source: '', deps: {} },
+    ];
+
+    expect(feedEntrypoint(feed).id).toBe(3);
+});
+
+test('determines entrypoint for feed with multiple entrypoints', async () => {
```
Comment (SimenB) [inline test/feed-entrypoint.test.js]: does this test make any sense? When do you have multiple entrypoints in a single feed?

File: test/feed-entrypoint.test.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
+'use strict';
```
Comment (SimenB) [inline test/feed-entrypoint.test.js]: None of the tests in this file needs `async`

File: package.json [hunk 14]
```diff
@@ -23,11 +26,18 @@
   },
   "license": "MIT",
   "dependencies": {
+    "asset-pipe-common": "^1.0.0-beta.6",
     "browser-pack": "6.0.3",
-    "get-stream": "^3.0.0"
+    "browserify": "^15.2.0",
+    "deep-equal": "^1.0.1",
```
Comment (SimenB) [inline package.json]: https://www.npmjs.com/package/fast-deep-equal

**Comment by digitalsadhu** [comment]
Created a little POC to see how hard it would be to switch out browserify for something else (in this case parcel) #74

File: package.json (hunk 10 above)
Comment (digitalsadhu) [reply package.json]: There is a task for it in jira as part of the asset-pipe cleanup task

File: lib/reader.js (hunk 8 above)
Comment (digitalsadhu) [reply lib/reader.js]: Yea, could do

File: lib/hash-feed.js (hunk 5 above)
Comment (digitalsadhu) [reply lib/hash-feed.js]: awww yea, will try

File: README.md (hunk 2 above)
Comment (digitalsadhu) [reply README.md]: Not sure, I think data format is more appropriate... maybe. "Data structure" elicits whiffs of maps along with rich notes of sets and arrays for me.

File: lib/hash-feed.js (hunk 5 above)
Comment (digitalsadhu) [reply lib/hash-feed.js]: no, not really. Or at least not in a way that cleans much up. Still need to push items into a stream and get the .hash property at the end.

File: package.json (hunk 14 above)
Comment (digitalsadhu) [reply package.json]: doesn't like this is being used anyway

File: test/feed-entrypoint.test.js (hunk 12 above)
Comment (digitalsadhu) [reply test/feed-entrypoint.test.js]: I don't think you can so yea, could probably remove.

File: lib/hash-feed.js (hunk 5 above)
Comment (SimenB) [reply lib/hash-feed.js]: yeah, but using `getStream` would allow us to drop error handling, as it's done for us. What if the readstream fails?

File: package.json (hunk 14 above)
Comment (SimenB) [reply package.json]: Even better
